Dockerfile

```diff
@@ -0,0 +1,40 @@
+FROM ubuntu:20.04 as builder
+
+# Setup timezone
+RUN echo 'Etc/UTC' > /etc/timezone \
+    && ln -s /usr/share/zoneinfo/Etc/UTC /etc/localtime \
+    && apt-get update && apt-get install -q -y tzdata \
+    && rm -rf /var/lib/apt/lists/*
+
+# Install build dependencies
+RUN apt-get update -y && apt-get install -y --no-install-recommends \
+    build-essential \
+    cmake \
+    && rm -rf /var/lib/apt/lists/*
+
+COPY / /agent_protocol_splitter
+WORKDIR /agent_protocol_splitter
+
+RUN mkdir build \
+    && cd build \
+    && cmake .. \
+    && make
+
+
+FROM busybox
+
+# Copy dependecies
+COPY --from=builder /lib/x86_64-linux-gnu/libc.so.6 /lib/x86_64-linux-gnu/
+COPY --from=builder /lib/x86_64-linux-gnu/libgcc_s.so.1 /lib/x86_64-linux-gnu/
+COPY --from=builder /lib/x86_64-linux-gnu/libm.so.6 /lib/x86_64-linux-gnu/
+COPY --from=builder /lib/x86_64-linux-gnu/libpthread.so.0 /lib/x86_64-linux-gnu/
+COPY --from=builder /lib/x86_64-linux-gnu/libstdc++.so.6 /lib/x86_64-linux-gnu/
+COPY --from=builder /lib64/ld-linux-x86-64.so.2 /lib64/
+
+COPY --from=builder /agent_protocol_splitter/build/protocol_splitter /
+COPY /entrypoint.sh /
+
+ENTRYPOINT ["/entrypoint.sh"]
+
+EXPOSE 14540/udp
```

Comment:
A minor nitpick (no need to fix) to keep in mind in the future: rarely-changing Docker instructions should be closer to top for improved caching and layer reuse.

`EXPOSE ...` here means that every time the code changes (`COPY --from=builder ...`), new image layers with expose instructions have to be built.

Example: expose last
----------------------------

<details>
<summary>Show</summary>

```
/tmp/dockertest
λ cat Dockerfile
FROM scratch

ADD date /date

EXPOSE 8080/tcp


/tmp/dockertest
λ echo $(date) > date && docker build -t test .
Sending build context to Docker daemon  3.072kB
Step 1/3 : FROM scratch
 --->
Step 2/3 : ADD date /date
 ---> 7cdf307f0f6d
Step 3/3 : EXPOSE 8080/tcp
 ---> Running in c48b45df3b9d
Removing intermediate container c48b45df3b9d
 ---> fbe1307569c5
Successfully built fbe1307569c5
Successfully tagged test:latest

/tmp/dockertest
λ echo $(date) > date && docker build -t test .
Sending build context to Docker daemon  3.072kB
Step 1/3 : FROM scratch
 --->
Step 2/3 : ADD date /date
 ---> 117bb067fae0
Step 3/3 : EXPOSE 8080/tcp
 ---> Running in 5d60579ef1fc
Removing intermediate container 5d60579ef1fc
 ---> 50126aaa54f2
Successfully built 50126aaa54f2
Successfully tagged test:latest
```
</details>

Example: expose before
-------------------------------

Observe:
- Expose layer now uses cache (` ---> Using cache`)
- Expose layer gets same ID on each invocation

Same layer ID means that Docker image transfers get done faster when redundant layers don't have to be transferred again every time a new version of the image gets published. Just the actually different layers need to be transferred.

<details>
<summary>Show</summary>

```
λ cat Dockerfile
FROM scratch

EXPOSE 8080/tcp

ADD date /date


/tmp/dockertest
λ echo $(date) > date && docker build -t test .
Sending build context to Docker daemon  3.072kB
Step 1/3 : FROM scratch
 --->
Step 2/3 : EXPOSE 8080/tcp
 ---> Running in af46eb388183
Removing intermediate container af46eb388183
 ---> f9b5c04ca0fd
Step 3/3 : ADD date /date
 ---> f2cd69027a97
Successfully built f2cd69027a97
Successfully tagged test:latest

/tmp/dockertest
λ echo $(date) > date && docker build -t test .
Sending build context to Docker daemon  3.072kB
Step 1/3 : FROM scratch
 --->
Step 2/3 : EXPOSE 8080/tcp
 ---> Using cache
 ---> f9b5c04ca0fd
Step 3/3 : ADD date /date
 ---> f637537a22d5
Successfully built f637537a22d5
Successfully tagged test:latest
```
</details>